Add providers with statuspage.io to operator config

jhadvig opened the pull request:
/assign @benjaminapetersen

benjaminapetersen commented on pkg/console/subresource/configmap/configmap.go:
```diff
@@ -155,6 +159,17 @@ func servingInfo() yaml.MapSlice {
 }
 
+func consoleProviders(providers operatorv1.ConsoleProviders) yaml.MapSlice {
+	if providers.Statuspage != nil {
```
This should be good for now. I imagine eventually we will need to `slice := yaml.MapSlice{ }`, and then conditionally append items one-by-one, but not necessary until we have other providers.

jhadvig commented:
@benjaminapetersen  comments addressed. PTAL

jhadvig commented:
/retest

benjaminapetersen commented on test/e2e/providers_test.go:
```diff
@@ -0,0 +1,120 @@
+)
+
+func setupProvidersTestCase(t *testing.T) *testframework.Clientset {
+	client := testframework.MustNewClientset(t, nil)
```
Mostly I'm mentioning this to ensure a `tearDown()` func puts the `config` resources back in a `vanilla` state, and that it `poll` the `operator` `config` for `Available:True, Progressing:False, Failing:False` to ensure we are ready to go on to the next test (avoid flakes).

@zherman0 was working on a `testframework` generic `poll`.

benjaminapetersen replied:
I don't think we have to hold this PR for it, but we will have to go back and update them when the func is ready.

benjaminapetersen commented:
/hold
I believe this has to wait for 4.2

benjaminapetersen commented:
but otherwise lgtm

spadgett commented:
> /hold
> I believe this has to wait for 4.2

master is already 4.2 :)

benjaminapetersen commented:
/hold cancel

> > /hold
> > I believe this has to wait for 4.2
>
> master is already 4.2 :)

neat :D

benjaminapetersen commented on pkg/console/subresource/configmap/configmap_test.go:
```diff
@@ -316,13 +342,18 @@ func TestNewYamlConfig(t *testing.T) {
+				},
 			},
-			want: finalCMDefaultOKD,
+			want: finalCMProviders,
```
So we have 1 test here, called `TestNewYamlConfig`, updating the `want` from `finalCMDefaultOKD` to `finalCMProviders` is confusing, IMHO.

Our table tests should probably have several cases:

```golang
tests := []struct {
     name string
     args args
     want string
}{ 
     {
          name: "Test NewYamlConfig() with defaults",
          args: args{
              ...
              // no StatusPage.io to ensure that a null value doesn't break our funcs
          },
          want: finalDefaultOKD  // should we inline?  
     },{
          name: "Test NewYamlConfig() with Status.io provider",
          args: args{ 
             ...
             // includes a StatusPage.io value and ensure output is as expected.
          },
          want: finalCMProviders  // should we inline? 
     },
}
```
See [route.TestIsAdmitted()](https://github.com/openshift/console-operator/blob/ed50a574a7d78bd562a5da21e72c2bb5dab331b8/pkg/console/subresource/route/route_test.go#L117), I think we have a couple tests that do a good job of testing the boundaries of funcs

benjaminapetersen commented on pkg/console/subresource/configmap/configmap.go:
```diff
@@ -155,6 +159,17 @@ func servingInfo() yaml.MapSlice {
 	}
 }
 
+func consoleProviders(providers operatorv1.ConsoleProviders) yaml.MapSlice {
```
This may not need testing if our main public func is well tested.  See my other test comment about adding more rows to the table test.

jhadvig commented:
@benjaminapetersen I've update the PR based on your comments.
Added extra top the `TestNewYamlConfig()` test and also added cleanup method to the e2e test.
Not sure if any additional tests are needed since there are only two cases:
- providers with statuspageID is set
- providers are not set

PTAL

benjaminapetersen commented on pkg/console/subresource/configmap/configmap_test.go:
```diff
@@ -302,27 +327,45 @@ func TestNewYamlConfig(t *testing.T) {
 			want: finalCMDefaultOKD,
 		},
+		{
+			name: "Test NewYamlConfig() with Statuspage.io provider",
```
Great, thanks!

benjaminapetersen commented:
/lgtm

Thanks!
